Also pass unwanted outputs to post-build-hook

yorickvP opened the pull request:
# Motivation
#6311 caused only wanted outputs to be passed to the post-build-hook, causing #6960.

# Context
<!-- Provide context. Reference open issues if available. -->
Fixes #6960 

For a long time, the post-build-hook was passed all built store paths. This changed, causing some breakage.

<!-- Non-trivial change: Briefly outline the implementation strategy. -->
- Add a test to see if unwanted outputs are passed to the post-build-hook.
- Collect wanted+unwanted outputs in a separate set and thread them through `DerivationGoal::checkPathValidity`, `DerivationGoal::registerOutputs`, `DerivationGoal::assertPathValidity`.



<!-- Invasive change: Discuss alternative designs or approaches you considered. -->

<!-- Large change: Provide instructions to reviewers how to read the diff. -->

# Checklist for maintainers

<!-- Contributors: please leave this as is -->

Maintainers: tick if completed or explain if not relevant

 - [ ] agreed on idea
 - [ ] agreed on implementation strategy
 - [ ] tests, as appropriate
   - functional tests - `tests/**.sh`
   - unit tests - `src/*/tests`
   - integration tests - `tests/nixos/*`
 - [ ] documentation in the manual
 - [ ] code and comments are self-explanatory
 - [ ] commit message explains why the change was made
 - [ ] new feature or bug fix: updated release notes

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/tweag-nix-dev-update-43/25185/1

Ericson2314 commented:
What does the `post-build-hook` if the paths are substituted instead of freshly built?

Yes, this fixes a practical problem, but if the semantics of the hook differ based on how the store object was obtained, I feel a bit uncomfortable about that.

ollie-etl commented:
Is this likely the underlying bug for my observation that only a single output from a multi-output derivation is uploaded to s3 cache? 
This _used_ to work.

post-hook is 
```sh
#!/bin/sh
set -eu
set -f # disable globbing
export IFS=' '
exec nix copy --to "s3://<store-uri>?endpoint=https://storage.googleapis.com" $OUT_PATHS
```

I'm using nix 2.8.1

yorickvP commented:
@Ericson2314 the post-build-hook was never called for substituted paths, and doing so wouldn't really fit common usage (uploading everything you build to a cache somewhere).


@ollie-etl the difference comes because, before this bug, a build of `a.dev` would cause the build hook to be invoked with the paths of `a.out`, `a.dev`. (even though building `a.out` is a side-effect).
With the bug, the post-build hook would only get `a.dev` and `a.out` would go unnoticed.

ollie-etl commented:
@yorickvP Here i'm building `a`, and `a.out` is being uploaded, but not `a.dev`

Ericson2314 commented:
@yorickvP BTW check out https://github.com/NixOS/nix/pull/6312 and in particular the settings `wantedOuputs = Outputs::All { };` I used to do (and Nix originally did). I think that might be a simpler way to achieve this PR too?

yorickvP commented:
Refactored it quite a bit, moved the `wantedOutputs` filtering to the end of the pipeline in `done()`.
CA tests fail as @thufschmitt predicted. Will look into it tomorrow.

Ericson2314 commented on src/libstore/outputs-spec.hh:
```diff
@@ -63,6 +67,9 @@ struct OutputsSpec : _OutputsSpecRaw {
     static std::optional<OutputsSpec> parseOpt(std::string_view s);
 
+    /* Filter a DrvOutputs by these output names */
+    DrvOutputs filter(DrvOutputs&& input) const;
```
I think this would be better as a method on `DrvOutputs` than `OutputsSpec`. `OutputsSpec` is the much simpler data type that shouldn't know anything about more complex things, even if the header just gets an abstract type.

yorickvP replied:
`DrvOutputs` is a `std::map`. Should I subclass or just write a separate function?

Ericson2314 replied:
@yorickvP Ah, I see. Either a standalone function or make a struct with one field then, I think.

yorickvP commented:
I spent a while debugging the test, but copying only the outputs doesn't seem to be sufficient for the CA test in any case (without my patch aswell).

@thufschmitt do you know if that's correct and supposed to be the case?

```
out=nix-build ./ca-thing.nix
nix copy --to cache $out
nix copy --from cache -f ca-thing.nix
    error: unable to build with a primary store that isn't a local store; either pass a different '--store' or enable remote builds.
```

thufschmitt commented:
> @thufschmitt do you know if that's correct and supposed to be the case?
> 
> ```
> out=nix-build ./ca-thing.nix
> nix copy --to cache $out
> nix copy --from cache -f ca-thing.nix
>     error: unable to build with a primary store that isn't a local store; either pass a different '--store' or enable remote builds.
> ```

Yes, it is. It's explained a bit more in https://www.tweag.io/blog/2021-02-17-derivation-outputs-and-output-paths/, but the gist of it is that with CA derivations there's a split between the output path and the information “this Derivation built this output path”. So if you want to be able to fetch “the output of this derivation” then you must copy exactly that (`nix copy --to cache -f ca-thing.nix` or `nix copy --to cache /nix/store/....-ca-thing.drv`) and not just the output path

yorickvP commented:
Okay, in that case this will have to suffice

alyssais commented:
Any news here?

yorickvP commented:
@thufschmitt rebased, please review

ollie-etl commented:
@thufschmitt @yorickvP Is this dead?

thufschmitt commented on tests/push-to-store-old.sh:
That's unrelated to this PR, but this file is weird :thinking:

thufschmitt reviewed (approved):
Sorry for taking so long at reaching here.

I'm still a bit sad this can't be a more local change, but it does fix the problem while making the code marginally more sensible. So let's roll!

thufschmitt commented:
@yorickvP can I let you rebase one last time? The conflicts seem mostly benign

thufschmitt reviewed (approved):
It's all passing, and looking good :rocket:

thufschmitt reviewed (approved):
It's all passing, and looking good :rocket:
